### ISSUE TITLE
Handle route relations for roads

### ISSUE THREAD

**Issue by pnorman** [geofabrik/shortbread-docs#49]
Highway refs are often indicated by route relations, so we need to support them.

One relation may be part of multiple road route relations, each of which has a ref and network tag. They also might be part of no road route relations and have a ref tag. I don't think we need to handle the third case of a ref tag plus refs from relations any different than the first tag.

**Comment by westnordost** [comment]
Bake the refs of the route relations onto the roads, right?

**Comment by pnorman** [comment]
Yes, we need the line strings to have appropriate properties. The difficult part is handling multiple relations.

**Comment by 1ec5** [comment]
Route relations intentionally provide more structure than way refs, so the two mechanisms aren’t quite equivalent – the structure should persist in the tile’s features to some extent, or it would defeat the purpose of route relations. openmaptiles/openmaptiles#1620 adopts a predictable `route_#_key` format for properties dissolved from route relations, allowing clients to process each value without resorting to any parsing.[^semicolon]

If you need to cap the maximum number of concurrent routes along a given way for performance reasons, here are some statistics to help you arrive at a practical limit: https://github.com/ZeLonewolf/openstreetmap-americana/issues/894#issuecomment-1765395845 https://github.com/ZeLonewolf/openstreetmap-americana/issues/894#issuecomment-1888362411.

[^semicolon]: This would also be a sound server-side solution to the problem of pretty-printing semicolon-delimited value lists.

**Comment by 1ec5** [comment]
> Yes, this is unfortunate. There's been a push towards the assumption that styles are going to do a bunch of country-specific route relation processing on the network tag. This is not practical as there is no global way to parse the network tag, and expecting general-purpose styles to have hundreds of rules for route relations isn't reasonable.

I don’t think ref formatting really needs to concern the specification, as the visual treatment will depend on the style. A topo map can likely get away with labeling Interstate&nbsp;5 as a plain “5”. A map that places slightly more emphasis on transportation probably should display graphical shields in regions that primarily use them to communicate the network. I recognize that this isn’t as easy as the traditional approach of slapping a tag value on a capsule, but it is a strong user expectation already supported by major parts of the OSM ecosystem.

Hopefully there isn’t any reason to parse either `network=*` or `ref=*`. In most regions where an alphabetic prefix is a core part of the route number, that prefix is already included in the route relation’s `ref=*`; a data consumer should not attempt to append one based on the `network=*`. Of the countries you listed in https://github.com/gravitystorm/openstreetmap-carto/issues/596#issuecomment-47328827, most are consistent about whether to include the prefix in the route `ref=*` based on the signs. Over the last couple years, OSM Americana contributors have worked with a number of countries to add or remove prefixes based on the signs, leaving a handful of prominent outliers like Germany, Brazil, and _maybe_ South Africa (depending on whether you consider them to still use shields or not).

> We can never expect implementations to pull down definitions at build time, and I don't think requiring 400+ formatting rules in the spec is reasonable.

Given the inherent complexity, Shortbread can reasonably choose not to support human-readable formatted route numbers based on route relation `ref=*`s. I normally don’t bring up Wikidata’s coverage of formatted road numbers in a style context, because they’re much more useful in other parts of a navigation system’s UI and in voice guidance instructions, for which generating definitions at build time is quite feasible.

Shortbread clients’ efforts would be better spent figuring out how to display graphical shields. For the Web, the [maplibre-shield-generator](https://www.npmjs.com/package/@americana/maplibre-shield-generator) plugin can draw shields based on raw `network=*` and `ref=*` values, covering a litany of edge cases.

> In Maplibre, most likely. Not in Mapnik.

Oh, I wasn’t aware that Shortbread is also targeting Mapnik styles. That could turn out to be an important constraint worth documenting somewhere.

**Comment by pnorman** [comment]
> If so, definitely consider putting `cycle_network=*` in a more generally named property

I hadn't considered handling different tags depending on the `route` value but I guess cycle network stuff is a bit different from anything else

**Comment by IloveRumania** [comment]
I'm not sure if this is relevant, but motorways without a `ref=*` tag don't display any label at all, [even if they have an `unsigned_ref=*` tag](https://osm.org/go/Ta0S5q4S?layers=S) [or either `name=*`, `noref=yes`, or both](https://osm.org/go/T2Jr406A-?layers=S). On the contrary, [trunk roads and below that have a `ref=*` tag but no `name=*` tag](https://osm.org/go/Tew1B0ng--?layers=S) also have no label. Can this be fixed? Thanks.

**Comment by pnorman** [comment]
> I'm not sure if this is relevant, but motorways without a `ref=*` tag don't display any label at all

The [street labels](https://shortbread-tiles.org/schema/1.0/#layer-street_labels) layer includes a name for all classifications. A particular style might decide not to render names for some classifications.

**Comment by IloveRumania** [comment]
Hm, this doesn't fix the issue. To repeat: Sometimes, when a value is absent where one is expected. In that case, the street's label should fall back to a label that already exists (e.g., `ref` and `name`).

**Comment by 1ec5** [comment]
Maybe what you’re noticing is the stylesheet’s [choice](https://github.com/versatiles-org/versatiles-style/blob/37dd0751452ebb364f3cece4b0dbb4ffccb2945f/src/shortbread/layers.ts#L433) not to label names on motorways? This repository is about the vector tile schema, not the stylesheet. Although the OSM website calls the layer “Shortbread”, it’s actually using a stylesheet called VersaTiles Colorful, which is maintained in [a different repository](https://github.com/versatiles-org/versatiles-style/issues/).

**Comment by pnorman** [comment]
Focusing on highways, I propose the following changes to the "street_labels" layer

-----

| Field name                 | Type   | Description |
|----------------------------|--------|-------------|
| `ref`                      | string | Value of OSM `ref` tag. If no ref tag is present, the `ref` tag from the most significant route relation the way is part of |
| `route_<TYPE>_<N>_ref`     | string | The `ref` tag for the N'th route relation of `<TYPE>`, only for highway=* |
| `route_<TYPE>_<N>_network` | string | The `ref` tag for the N'th route relation of `<TYPE>`, only for highway=*  |

### Route relations
Route relations are split by the value of the `type` tag and that have no `ref` tag and no `network` tag are removed. Within each `type` the relations are sorted by significance in a manner up to implementations. Each route of  type `road`, `bicycle`, and `foot` on a `highway` way is included.

Implementations may have a maximum number of route relations they will handle or a maximum number per type.

------
Routes for `railway=*` belong in a different issue. I'm not dead set on the list of route types we have here but my thoughts are
1. Road relations are essential to general purpose maps in locations with many route signs and concurrencies because they are so important for orientation
2. Cycle and foot relations are not as essential for general purpose maps but are still useful in many areas.
3. hiking and mtb routes are more the realm of topographic maps which require many features not present in a light-weight schema
4. A map that wants to show bus, tram, trollybus, share_taxi, etc  will require a lot more transit-related information that a light-weight schema lacks.

**Comment by 1ec5** [comment]
I agree with including cycling route relations. Conventional road atlases include long-distance cycling routes because cyclists use the road too. Sometimes these maps also mark long-distance hiking routes, but not as often. Leaving out public transportation routes is a reasonable move for now, though I could see them becoming more desirable over time as people more details styles from these tiles.

Would `route_bicycle_#_network` be based on `network=*cn` or `cycle_network=*`? The three-letter codes can be useful for choosing the zoom levels to draw a route line at, but that isn’t the purpose of the `street_labels` layer. `cycle_network=*` is much more useful for drawing shields. (It’s a real shame that OSM can’t swap the two keys at this point.)

Some route networks (especially toll and scenic route networks) rely on something other than a number inside the shield to distinguish the various routes in the network, such as a color or a pictogram. These routes won’t contain `ref=*`, but they might have distinct `ref:colour=*` or `wikidata=*` tags. By omitting these routes, we’re counting on the member roadways to have a `name=*` that the style can label instead.

The new attributes are limited to `highway=*`, but some road and cycling routes follow ferry lines (`route=ferry`) out of necessity. It’s rare, but naturally these cases are usually important long-distance routes.

**Comment by pnorman** [comment]
The discussion at SOTM with cycling experts was to use `network`.



> The new attributes are limited to `highway=*`, but some road and cycling routes follow ferry lines (`route=ferry`) out of necessity. It’s rare, but naturally these cases are usually important long-distance routes.

Do you propose extending this proposal to the `ferries` layer?

**Comment by 1ec5** [comment]
> The discussion at SOTM with cycling experts was to use `network`.

In that case, a style that marks cycling routes with shields will need a separate `route_bicycle_#_cycle_network` property. Maybe we can think of something that doesn’t sound so redundant. This use case is probably more prominent in some countries than others. For example, `network=*cn` is useless for shields in most of North America because it’s too generic. Recently someone tried to [fork OSM Americana](https://github.com/kaneap/openstreetmap-americana-bike/) to mark bike routes, but they ended up only being able to support one kind of national route in the U.S., and one kind of regional route in one county in Wisconsin. The map is only useful in that one county and can’t scale beyond that. OpenMapTiles doesn’t geocode every feature, so their `ncn` (USBR) and `rcn` (Milwaukee County) shields appear incongruously worldwide.

A more obvious main use case for `network=*cn` is to color-code route lines by (more or less) geographic scope, or to determine the minimum zoom level – similar to the use cases for `highway=*`. Will that be the intended use of this layer? If so, there might be interest in something analogous for road routes too, but it wouldn’t come from those relations’ `network=*` tags verbatim. For a while, OpenMapTiles tried to maintain a similarly generalized classification, but they gave up because it backed the schema into a backwards compatibility corner: openmaptiles/openmaptiles#1648.

Another solution would be to convince the North American communities to ditch `network=*cn` in favor of what’s currently in `cycle_network=*`, but that seems very far-fetched at the moment. I haven’t even managed to get `state=proposed` deprecated yet!

> Do you propose extending this proposal to the `ferries` layer?

Are you still thinking of copying the route-related attributes onto features in the `streets` layer? If so, the `ferries` layer would need the same attributes. Unfortunately, style designers would have to duplicate any shield-related styling on another style layer for another tile layer. Pretty annoying, and not likely to happen.

Or would you take the same approach as OpenMapTiles with redundant geometries in a separate layer? Their `transportation_name` layer only contains roadway names and route-related metadata, so a feature in this layer can theoretically coincide with a ferry route. However, they seem to have a bug that suppresses the ferry segments of those routes: osm-americana/openstreetmap-americana#488.

**Comment by pnorman** [comment]
> Are you still thinking of copying the route-related attributes onto features in the `streets` layer?

What I proposed and we've been discussing is [above](https://github.com/shortbread-tiles/shortbread-docs/issues/49#issuecomment-5388952329)

My argument against being worried about cycle_network is that that information isn't suited to a light-weight general purpose schema but better belongs to either something focused on cycle routes or something heavier weight that tries to give enough information to distinguish all route symbols. To a certain extent that applies to road route relations too.

**Comment by 1ec5** [comment]
I guess the part I’m unclear about is what a Shortbread-based style could possibly use the cycling routes for. What did the folks at SotM suggest they’d use it for?

https://github.com/shortbread-tiles/shortbread-docs/issues/49#issuecomment-5388952329 alludes to the routes being used in non-specialized maps. I figure that any map that depicts both highway routes and cycling routes would need to visually distinguish them. `network=*cn` and `ref=*` alone are only enough to achieve something like OpenCycleMap, with its highlighter-like route overlay and plain numbers color-coded by `network=*`. You’re already enabling something more sophisticated for road routes.

Many transportation maps in my collection since the 2000s feature some pattern-accurate bike route shields, but only one color-codes bike route lines. That’s not entirely surprising since I mostly collect North American maps, which don’t color-code lines much. But do European transportation maps commonly highlight bike routes in that manner?

Introducing a `route_bicycle_#_network` attribute commits us to the `*cn` scheme from now until v2.0 and probably precludes anything more sophisticated until then, unless we accept a redundant-sounding attribute name.

**Comment by pnorman** [comment]
> I guess the part I’m unclear about is what a Shortbread-based style could possibly use the cycling routes for. What did the folks at SotM suggest they’d use it for?

@systemed Could you speak to using network from cycle route relations? You're more experienced with them.

-----

Are we in general agreement on route=road relations? If so I'd like to get a PR for that.
